### PR TITLE
Reserve room detail

### PR DESCRIPTION
**Pull request by owen-le-goldenowl**
# Issue

- Build detail screen of booked room.

# Description

- Fix error if user has not upload avatar.
- Change database schema:

    1. Delete _many-to-many_ association between **orders** and **services** table.
    2. Add _many-to-many_ association between **bookings** and **services** table.
    3. Change _has-many_ **bookings** association of the **orders** table to belongs-to **booking**.
- Cancel a booked room.
- Check in a booked room.

# Pictures

### CODE REVIEW

File: app/controllers/bookings_controller.rb [hunk 1]
```diff
@@ -31,22 +32,48 @@ def create
...
+      customer_nin: @booking.customer.customer_nin,
+      phone_number: @booking.customer.phone_number,
+      men: @booking.guests.where(guest_type: 'men').first&.quantity || 0,
+      women: @booking.guests.where(guest_type: 'women').first&.quantity || 0,
+      baby_girl: @booking.guests.where(guest_type: 'baby_girl').first&.quantity || 0,
+      baby_boy: @booking.guests.where(guest_type: 'baby_boy').first&.quantity || 0,
+      services: Service.all.map { |s| [s.name, (@booking.service_uses.where(service_id: s.id).first&.quantity || 0), s.price]}
+    )
+    # puts @room_booking.inspect
+    puts @room_booking.inspect
```
Comment (tycenguyen) [inline app/controllers/bookings_controller.rb]: delete this code

File: app/controllers/bookings_controller.rb [hunk 2]
```diff
@@ -76,6 +101,77 @@ def update
...
+      name: @booking.customer.name,
+      customer_nin: @booking.customer.customer_nin,
+      phone_number: @booking.customer.phone_number,
+      men: @booking.guests.where(guest_type: 'men').first&.quantity || 0,
+      women: @booking.guests.where(guest_type: 'women').first&.quantity || 0,
+      baby_girl: @booking.guests.where(guest_type: 'baby_girl').first&.quantity || 0,
+      baby_boy: @booking.guests.where(guest_type: 'baby_boy').first&.quantity || 0,
+      # format: [[service_name, quantity, price]]
+      services: Service.all.map { |s| [s.name, @booking.service_uses.where(service_id: s.id).first&.quantity || 0, s.price]}
+    )
```
Comment (tycenguyen) [inline app/controllers/bookings_controller.rb]: separate this object build to private method to re-use

File: app/controllers/bookings_controller.rb [hunk 3]
```diff
@@ -76,6 +101,77 @@ def update
...
+
+    if params['cancel_booking']
+      destroy
+    elsif params['commit']
+
+      if @booking.update!(status: 'checked_in')
```
Comment (tycenguyen) [inline app/controllers/bookings_controller.rb]: update then check valid?, update! do not return false when failed

File: app/controllers/bookings_controller.rb [hunk 4]
```diff
@@ -76,6 +101,77 @@ def update
...
+        end
+        # only update if user change guest quantity
+        if @room_booking.baby_girl != baby_girl_count
+          @booking.guests.baby_girl.first_or_initialize.update(quantity: baby_girl_count)
+        end
+        # only update if user change guest quantity
+        if @room_booking.baby_boy != baby_boy_count
+          @booking.guests.baby_boy.first_or_initialize.update(quantity: baby_boy_count)
+        end
+
```
Comment (tycenguyen) [inline app/controllers/bookings_controller.rb]: Wrap each update by assign attribute to object and save at last

File: app/controllers/bookings_controller.rb [hunk 5]
```diff
@@ -76,6 +101,77 @@ def update
...
+        # update service uses
+        @room_booking.services.each do |s|
+          service_name = s[0]
+          old_quantity = s[1].to_i
+          new_quantity = params[s[0]].to_i
+
+          # only update if any servie input value changes
+          if old_quantity != new_quantity
+            service_record = Service.find_by_name(service_name)
+            @booking.service_uses.create!(service_id: service_record.id, quantity: new_quantity)
```
Comment (tycenguyen) [inline app/controllers/bookings_controller.rb]: don't use create!, exception will stop the flow

File: app/controllers/bookings_controller.rb [hunk 6]
```diff
@@ -76,6 +101,77 @@ def update
...
+          end
+        end
+
+        redirect_to rooms_path(hotel_id: Room.find(params[:room_id]).hotel.id)
```
Comment (tycenguyen) [inline app/controllers/bookings_controller.rb]: re-check the object relationship, no need to query find

File: app/controllers/bookings_controller.rb [hunk 7]
```diff
@@ -76,6 +101,77 @@ def update
...
+        render :show
+      end
+    end
+  end
+
+  def destroy
+    booking= Booking.find(params[:id])
+    room_id =  booking.room_id
+    booking.destroy
+    redirect_to rooms_path(hotel_id: Room.find(room_id).hotel.id)
```
Comment (tycenguyen) [inline app/controllers/bookings_controller.rb]: re-check the object relationship

File: app/controllers/rooms_controller.rb [hunk 8]
```diff
@@ -7,25 +7,32 @@ def index
     @checkin_time = params[:checkin_time] || Time.current.change(hour: 12)
     @checkout_time = params[:checkout_time] || 1.day.from_now.change(hour: 12)
 
-    @available_rooms_groups = available_rooms.group(:room_type).count # example: {"queen"=>10, "double"=>10, "single"=>10}
+    # example: {"queen"=>10, "double"=>10, "single"=>10}
+    @available_rooms_groups = available_rooms.group(:room_type).count
```
Comment (tycenguyen) [inline app/controllers/rooms_controller.rb]: use size instead count

File: app/forms/room_booking.rb [hunk 9]
```diff
@@ -1,13 +1,14 @@
...
+  attr_accessor :name, :customer_nin, :phone_number, :rent_type, :room_id, :checkin_date, :checkin_time, :men, :women, :baby_girl, :baby_boy, :booking_id, :services
 
   include ActiveModel::Model
   extend ActiveModel::Naming
   include ActiveModel::Conversion
   include ActiveModel::Validations
 
   def initialize(params)
+    @room_id = params.fetch(:room_id, Room.new.id)
+    @booking_id = params.fetch(:booking_id, Booking.new.id)
```
Comment (tycenguyen) [inline app/forms/room_booking.rb]: remove [object].new.id

File: app/controllers/bookings_controller.rb [hunk 10]
```diff
@@ -166,14 +113,32 @@ def check_in
   end
 
   def destroy
-    booking= Booking.find(params[:id])
-    room_id =  booking.room_id
+    booking = Booking.find(params[:id])
```
Comment (tycenguyen) [inline app/controllers/bookings_controller.rb]: create hotel variable to store id

Comment (owen-le-goldenowl) [reply app/controllers/bookings_controller.rb]: 41e91c95dcaa36adfc66d993e2bbabb66332e776
@tycenguyen

**Pull request merged**